pkg/planner/core/partition_prune.go

```diff
@@ -38,6 +38,48 @@ func PartitionPruning(ctx base.PlanContext, tbl table.PartitionedTable, conds []
 			return nil, err
 		}
 		ret := s.convertToIntSlice(rangeOr, pi, partitionNames)
+		if pi.CanHaveOverlappingDroppingPartition() {
+			if len(ret) == 1 && ret[0] == FullRange {
+				ret = make([]int, 0, len(pi.Definitions))
+				for i := range pi.Definitions {
+					ret = append(ret, i)
+				}
+			}
+			newRet := make([]int, 0, len(ret))
+			for i := range ret {
+				idx := pi.GetOverlappingDroppingPartitionIdx(ret[i])
+				if idx == -1 {
+					// dropped without overlapping partition, skip it
+					continue
+				}
+				if idx == ret[i] {
+					// non-dropped partition
+					newRet = append(newRet, idx)
+					continue
+				}
+				// partition being dropped, remove the consecutive range of dropping partitions
+				// and add the overlapping partition.
+				end := i + 1
+				for ; end < len(ret) && ret[end] < idx; end++ {
+					continue
+				}
+				// add the overlapping partition, if not already included
+				if end >= len(ret) || ret[end] != idx {
+					// It must also match partitionNames if explicitly given
+					if len(partitionNames) == 0 || s.findByName(partitionNames, pi.Definitions[idx].Name.L) {
+						newRet = append(newRet, idx)
+					}
+				}
+				if end < len(ret) {
+					newRet = append(newRet, ret[end:]...)
+				}
+				break
```

Comment:
Emm this part seems tricky, could you please add a concrete example here?

Reply:
Yes it is a bit tricky, not sure how to express the code in the best way.
But what it should do is:
from a list of indexes/offsets of pi.Definitions (i.e partitions to be used after partition pruning)
- if full range, replace the 'full range' marker with a range of all partitions
- remove the partitions that are being dropped
- if not already included add the 'overlapping' partition (also check explicit partition pruning in partitionNames)

How it is implemented is by looping over all used partitions to create a new set of used partitions
- add the non-dropping partitions
- Until an 'overlapping' partition is found, then
  - skip the full range of dropping partitions (for range partitioning, one can only drop a consecutive set of partitions).
  - if not already included, add the 'overlapping' partition, with a check that it also matches explicit partition pruning/partitionNames if given.
  - append all following used partitions.
And before returning, check if it matches the full range, if so return the simpler FullRange marker.

Actual example:
```sql
CREATE TABLE t (a int, b varchar(255))
PARTITION BY RANGE (a)
(PARTITION p0 VALUES LESS THAN (10),
 PARTITION p1 VALUES LESS THAN (20),
 PARTITION p2 VALUES LESS THAN (30),
 PARTITION p3 VALUES LESS THAN (40));
-- various DML...
ALTER TABLE t DROP PARTITION p1,p2;
```

During DDL state StateWriteOnly, the dropping partitions will still be there, but 'blocked' for read and write, so for write it will simply give an error, but for read we must take care for the possibility that other sessions might see the next DDL state StateDeleteOnly, where the partition p1 and p2 are gone, so p3 will accept values into the same range as p1 and p2 previously had.

So to not hide any values or have inconsistency depending on if you do pruning or not we need to replace dropping partitions with the 'overlapping' partition when doing partition pruning (filtering etc. will then make sure that we still return the matching rows.

@qw4990 do you want me to go through the code what happens in this scenario?